### ISSUE TITLE
device_cgroup_rules misbehave under v2.2.2

### ISSUE THREAD

**Issue by Paraphraser** [docker/compose#9059]
**Description**

I suspect there may be a problem in how docker-compose v2.2.2 implements `device_cgroup_rules`. On a system where the **ONLY** difference is the version of docker-compose being used:

* docker-compose 1.29.2 – succeeds
* docker-compose v2.2.2 – fails

**Steps to reproduce the issue:**

This problem may be reducible to a simple test case but I'm afraid that I simply don't have the requisite knowledge to attempt it. I can only report what I see with the tools at my disposal.

The host system is a Raspberry Pi 3B+ running Raspbian Buster which is fully up-to-date with `apt upgrade`, plus Docker, plus [octoprint-docker](https://github.com/OctoPrint/octoprint-docker) running in a container.

> I can't use Raspbian Bullseye because camera support appears to be broken. I am not sure whether this is an OS, OctoPrint or octoprint-docker problem. I am stuck with Buster until the camera issue is resolved.

This is the basic setup:

1. A Docker container, running a modified version of [octoprint-docker](https://github.com/OctoPrint/octoprint-docker), is defined as follows:

	```yaml
	octoprint:
	  container_name: octoprint
	  build: ./.templates/octoprint/.
	  restart: unless-stopped
	  environment:
	  - TZ=Australia/Sydney
	  - ENABLE_MJPG_STREAMER=true
	  - MJPG_STREAMER_INPUT=-r 1152x648 -f 10
	  - CAMERA_DEV=/dev/video0
	  ports:
	  - "9980:80"
	  device_cgroup_rules:
	  - "c 188:* rw"
	  - "c 81:* rw"
	  volumes:
	  - ./volumes/octoprint:/octoprint
	  - /dev:/host/dev:ro
	```

	Two files are added to the container via Dockerfile:
	
	- a "reaction script" (discussed below); and
	- a slightly modified version of an existing s6 script associated with camera support.
	
	If it will aid diagnosis, full documentation of the arrangement can be found at [IOTstack+OctoPrint: When your 3D printer turns on and off](https://gist.github.com/Paraphraser/650eeab26794dc7363b7123b33f671c4).
	
2. Assume the container is up but no 3D printer is connected.
3. A 3D printer is powered-on and becomes visible on a USB port.
4. A udev rule matches on the device parameters and invokes `docker exec` to trigger a "reaction script" inside the container, passing the printer's device name.
5. Via the volume mapping `/dev:/host/dev:ro`, the reaction script can see the host's `/dev`. The script conforms the **internal** `/dev` to match the host's, and ends by restarting the OctoPrint process running within the container. The script can do this because of the wildcards in:

	```yaml
	device_cgroup_rules:
	- "c 188:* rw"
	- "c 81:* rw"
	```
	
	The first rule is for printers, the second for cameras.

6. Once the restart is complete, the OctoPrint process can see the printer in the container's `/dev`, and will present it in the GUI, after which jobs can be sent to the printer.

A similar chain of events occurs when the printer is switched off. The udev rule triggers the reaction script which removes the internal devices and restarts the OctoPrint process.

This scheme has been working well for the better part of a year. It is robust and reliable.

**Describe the results you expected - docker-compose 1.29.2:**

Assume:

1. No 3D printer is connected to the Raspberry Pi.
2. The stack is down.
3. The system has just been rebooted.

> Rebooting is not actually necessary - it just makes the point that there is nothing held over from prior testing.

I start the stack with docker-compose 1.29.2:

```bash
$ cd ~/IOTstack

$ docker-compose version | head -1
docker-compose version 1.29.2, build unknown

$ docker-compose up -d 
Creating network "iotstack_default" with driver "bridge"
Creating portainer-ce ... done
Creating octoprint    ... done
```

Immediately after starting, the container has no knowledge of either the 3D printer or the video camera:

```bash
$ docker exec octoprint ls -l /dev/AnyCubicI3Mega /dev/video0
ls: cannot access '/dev/AnyCubicI3Mega': No such file or directory
ls: cannot access '/dev/video0': No such file or directory
```

I connect the printer. After the udev rule has fired and the reaction script has run, the container knows about both the printer and the video camera:

```bash
$ docker exec octoprint ls -l /dev/AnyCubicI3Mega /dev/video0
crw-r--r-- 1 root root 188, 0 Dec 21 17:21 /dev/AnyCubicI3Mega
crw-r--r-- 1 root root  81, 4 Dec 21 17:21 /dev/video0
```

A (desirable) side-effect of the container starting to communicate with the Raspberry Pi's "ribbon-cable" camera is the indicator LED on the camera PCB turning on. As soon as the LED illuminates, I can connect to the camera stream URL independently of the OctoPrint UI.

If I open the OctoPrint web GUI, select the printer and click "connect", the OctoPrint terminal log shows:

```
Changing monitoring state from "Offline" to "Opening serial connection"
Connecting to port /dev/AnyCubicI3Mega, baudrate 250000
Changing monitoring state from "Opening serial connection" to "Connecting"
Connected to: Serial<id=0x6b3b1478, open=True>(port='/dev/AnyCubicI3Mega', baudrate=250000, bytesize=8, parity='N', stopbits=1, timeout=10.0, xonxoff=False, rtscts=False, dsrdtr=False), starting monitor
Send: N0 M110 N0*125
Recv: start
Send: N0 M110 N0*125
Recv: echo:V1.1.2
Recv:  1.1.0-RC8

    … many more lines … eventually settles to repeating

Send: M105
Recv: ok T:26.0 /0.0 B:27.7 /0.0 @:0 B@:0
```

It is clear that OctoPrint is communicating with the printer. 
Everything in the above is the expected result.

To end the test, I disconnect the printer, "down" the stack and reboot the Pi.

**Describe the results you received - docker-compose v2.2.2:**

Assume:

1. No 3D printer is connected to the Raspberry Pi.
2. The stack is down.
3. The system has just been rebooted.

> Rebooting is not necessary - it just makes the point that there is nothing held over from prior testing.

I start the stack with docker-compose v2.2.2 (invoking it using plugin syntax):

```bash
$ cd ~/IOTstack

$ docker compose version
Docker Compose version v2.2.2

$ docker compose up -d 
$ docker compose up -d
[+] Running 3/3
 ⠿ Network iotstack_default  Created                                                                                   0.5s
 ⠿ Container portainer-ce    Started                                                                                   3.2s
 ⠿ Container octoprint       Started                                                                                   3.2s
```

Immediately after starting, the container has no knowledge of either the 3D printer or the video camera:

```bash
$ docker exec octoprint ls -l /dev/AnyCubicI3Mega /dev/video0
ls: cannot access '/dev/AnyCubicI3Mega': No such file or directory
ls: cannot access '/dev/video0': No such file or directory
```

I connect the printer. After the udev rule has fired and the reaction script has run, the container knows about both the printer and the video camera:

```bash
$ docker exec octoprint ls -l /dev/AnyCubicI3Mega /dev/video0
crw-r--r-- 1 root root 188, 0 Dec 21 17:29 /dev/AnyCubicI3Mega
crw-r--r-- 1 root root  81, 0 Dec 21 17:29 /dev/video0
```

Up until this point, everything has been the same as when I use version 1.29.2 of docker-compose. This is where things diverge.

Firstly, the camera side-effect mentioned before does not occur. The PCB indicator LED does not turn on and I am not able to connect to the camera stream URL. After a timeout, the browser responds with "503 Service Unavailable".

Secondly, if I open the web GUI, the printer is visible in the popup menu as expected. But when I select the printer and click "connect", the UI raises an error and its log shows:

```
Changing monitoring state from "Offline" to "Opening serial connection"
Connecting to port /dev/AnyCubicI3Mega, baudrate 250000
Changing monitoring state from "Opening serial connection" to "Error"
Changing monitoring state from "Error" to "Offline after error"
Unexpected error while connecting to serial port /dev/AnyCubicI3Mega, baudrate 250000 from hook default: SerialException: '[Errno 1] could not open port /dev/AnyCubicI3Mega: [Errno 1] Operation not permitted: '/dev/AnyCubicI3Mega'' @ comm.py:_open_serial:3775
```

Both the camera and printer comms are failing. What they appear to have in common is `device_cgroup_rules`.

So, to boil it down to a nutshell, when all other factors are controlled and the version of docker-compose is the sole determinant:

* docker-compose 1.29.2 = works
* docker-compose v2.2.2 = does not work

**Output of `docker compose version`:**

Two versions:

1. 1.29.2:

	```bash
	$ docker-compose version
	docker-compose version 1.29.2, build unknown
	docker-py version: 5.0.0
	CPython version: 3.7.3
	OpenSSL version: OpenSSL 1.1.1d  10 Sep 2019
	```

2. v2.2.2:

	```bash
	$ docker compose version
	Docker Compose version v2.2.2
	```

**Output of `docker info`:**

```bash
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.6.3-docker)
  compose: Docker Compose (Docker Inc., v2.2.2)

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 7
 Server Version: 20.10.11
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.63-v7+
 Operating System: Raspbian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: armv7l
 CPUs: 4
 Total Memory: 871.7MiB
 Name: octopi
 ID: HHEC:Q244:H2SP:VKOT:T4C2:3KSC:WKH7:4OXS:7TQ2:E3N2:64XL:7OBX
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory TCP limit support
WARNING: No oom kill disable support
WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

**Additional environment details:**

```bash
$ cat /proc/device-tree/model
Raspberry Pi 3 Model B Plus Rev 1.3

$ uname -a
Linux octopi 5.10.63-v7+ #1496 SMP Wed Dec 1 15:58:11 GMT 2021 armv7l GNU/Linux

$ echo "COMPOSE_DOCKER_CLI_BUILD=$COMPOSE_DOCKER_CLI_BUILD, DOCKER_BUILDKIT=$DOCKER_BUILDKIT"
COMPOSE_DOCKER_CLI_BUILD=1, DOCKER_BUILDKIT=1
```

docker-compose 1.29.2 is installed at:

```
/usr/local/bin/docker-compose
```

docker-compose v2.2.2 is installed in two places:

```
~/.docker/cli-plugins/docker-compose
/usr/local/bin/docker-compose.v2.2.2
```

It makes no difference to the outcome of these tests whether I run v2.2.2 as:

```
$ docker-compose.v2.2.2 up -d
$ docker compose up -d
```

**Additional test:**

To try to confirm that `device_cgroup_rules` was more likely to be the culprit than something generic to device mapping under docker-compose v2.2.2, I tried replacing the rules with explicit `devices`:

```yaml
octoprint:
  container_name: octoprint
  image: octoprint/octoprint:latest
  restart: unless-stopped
  environment:
  - TZ=Australia/Sydney
  - ENABLE_MJPG_STREAMER=true
  - MJPG_STREAMER_INPUT=-r 1152x648 -f 10
  - CAMERA_DEV=/dev/video0
  ports:
  - "9980:80"
  devices:
  - /dev/AnyCubicI3Mega:/dev/AnyCubicI3Mega
  - /dev/video0:/dev/video0
  volumes:
  - ./volumes/octoprint:/octoprint
```

To use this pattern, the 3D printer must be turned on before the container is started, otherwise the 3D printer mapping will fail. So:

1. Stack down.
2. Printer turned on.
3. The udev rule fires and creates /dev/AnyCubicI3Mega on the host but its attempts to trigger the reaction script fail because:

	- the container is not yet running; and
	- even if the container was running, the `image` statement points to the unmodified image from DockerHub which wasn't built with the reaction routine.

	In other words, the only way the container can learn about the printer or camera is via the `devices:` mappings.
	
4. Bring up the stack with docker-compose v2.2.2. The camera LED turns on and the OctoPrint GUI shows the printer.
5. The printer can be selected, the "Connect" button clicked and OctoPrint connects to the printer successfully.

The logical conclusion is that there is nothing intrinsic to the setup which prevents the container from communicating with the printer when docker-compose v2.2.2 is used to bring up the stack. The problem "follows the `device_cgroup_rules`".

**Comment by Paraphraser** [comment]
I want to confirm that this is still a problem with versions 2.2.3 and 2.3.0.

**Comment by ndeloof** [comment]
can you give https://github.com/docker/compose/pull/9251 a try?
(run `make compose-plugin` to build from sources)

**Comment by Paraphraser** [comment]
I'm happy to do that but I'm not sufficiently familiar with the build process that I understand your instructions. I have previously done a `git clone` followed by a make. That worked and produced a binary but the how-to of applying #9251 to that basic process is something I just don't know how to do. I'd really appreciate some extra hints on this. Sorry to be a pain.

**Comment by KevinRohn** [comment]
Hey guys, 

just now, I was opening a new issue, but I`ve found this one is also existing. 
I spend some time writing this issue, so I post it there as a comment. Maybe it helps.

## Description

`device_cgroup_rules` are not working with `docker-compose` version `>=v2.0.0`.
 I spent hours to get hotplug of serial port working on my raspberry pi 3.

With `docker-compose` version `v1.29.2` everything works as expected.

## Steps to reproduce the issue:

**Setup`udev` rule, which reacts on _add,change and remove_ events of an external serial device.**

```
SUBSYSTEM=="tty", KERNEL=="ttyUSB*", KERNELS=="1-1.3:1.0", ATTRS{interface}=="FT232R USB UART", SYMLINK+="ttyPortBLE"

ACTION=="add|change", KERNEL=="ttyUSB*", KERNELS=="1-1.3:1.0", ATTRS{interface}=="FT232R USB UART", RUN+="/home/powerio/scripts/update_docker_tty.sh 'added' '/dev/ttyPortBLE' '%M' '%m'"
ACTION=="remove", ENV{ID_MODEL}=="FT232R_USB_UART", RUN+="/home/powerio/scripts/update_docker_tty.sh 'removed' '/dev/ttyPortBLE' '%M' '%m'"
```

**Script which is triggered by the `udev`. `update_docker_tty.sh`**
```
#!/bin/bash

# Script is called by udev events on add/change/or remove events.
# It maps/removes the device during the runtime to the container (Hot plug).
#
# Parameters:
# $1 event type : [ 'add','change','remove' ]
# $2 name of the tty dev : e.g. /dev/ttyPortBLE
# $3 group id: %M
# $4 system order number (starting from 0): %m

LOG=/var/log/update_docker_tty.log

DOCKER_CONTAINER="dev-container_dev-container-ble-ws-to-socket_1"

echo "tty event detected: $1 | $2 | $3 | $4" | tee -a $LOG
if [ ! -z "$(docker ps -qf name=$DOCKER_CONTAINER)" ]; then
    if [ "$1" == "added" ]; then # added
        if [ "$3" == "0" ]; then # added
            echo "Double add SKIP! $2" | tee -a $LOG
        else
            docker exec $DOCKER_CONTAINER rm -rf $2
            docker exec $DOCKER_CONTAINER mknod $2 c $3 $4
            #docker exec $DOCKER_CONTAINER chmod -R 660 $2
            docker exec $DOCKER_CONTAINER chgrp dialout $2
            echo "Added $2 to docker container $DOCKER_CONTAINER" | tee -a $LOG
        fi
    else
        # removed
        docker exec $DOCKER_CONTAINER rm -rf $2
        echo "Removed $2 from docker" | tee -a $LOG
    fi
else
    echo "Couldn't react to tty event. Docker container $DOCKER_CONTAINER is not running." | tee -a $LOG
fi
```

**Log output of the `udev` rule triggered script to verify the hot plug process**
```
Added /dev/ttyPortBLE to docker container dev-container_dev-container-ble-ws-to-socket_1
tty event detected: removed | /dev/ttyPortBLE | 188 | 4
Removed /dev/ttyPortBLE from docker
```

**The `docker-compose.yml` file with the `device_cgroup_rules` entry**
```
version: "3.7"
services:
  dev-container-ble-ws-to-socket:
    build: 
        context: ../
        dockerfile: dev-container/Dockerfile
    #privileged: true
    network_mode: host
    volumes:
      #- "/dev:/dev"
      - "../:/opt/ble-ws-to-socket"
      - /opt/ble-ws-to-socket/node_modules/
    device_cgroup_rules:
      - 'c 188:* rmw'
```

##Result with `v1.29.2`:
The docker container contains a Node.js application, which tries to open the serial port `/dev/ttyPortBLE` every second. As seen in the video, the port is unplugged on startup, after inserting the serial port to the device, the serial port is mapped to the docker container as expected and the port can be accessed.

https://user-images.githubusercontent.com/25847098/157433144-2dd454b1-8ab1-4bd6-ac60-012517fa41af.mov

##Result with `v2.2.3`:

Same setup as above. There are some small changes like:

In version `v1.29.2` the docker container name is: `dev-container-dev-container-ble-ws-to-socket_1` (underline)
In version `v2.2.3` the docker container name is:  `dev-container-dev-container-ble-ws-to-socket-1`

I also change the `DOCKER_CONTAINER` in `update_docker_tty.sh` variable to the matching docker container name.

The device `/dev/ttyPortBLE` is mapped to the container, but it's not accessible. I alos tried to add `docker exec $DOCKER_CONTAINER chmod -R 660 $2` and `docker exec $DOCKER_CONTAINER chgrp dialout $2` to the script. But this makes no difference.

https://user-images.githubusercontent.com/25847098/157433634-1f8b2033-938c-435d-8b98-19054bc897c9.mov

**Comment by Paraphraser** [comment]
TL;DR - it works!

The approach I used to testing #9251:

```
$ cd ~/compose
$ git pull
$ git fetch origin pull/9251/head:PR9251
$ git checkout PR9251
$ make
$ ./bin/docker-compose version
Docker Compose version v2.3.2-4-gd8775c7a
```

Test compose file:

```yaml
services:

  as-rule:
    container_name: as-rule
    image: ubuntu
    command: /bin/bash -c 'while ((1==1)) ; do sleep 1 ; done'
    device_cgroup_rules:
    - "c 81:* rw"
```

Three tests:

1. Test 1 = v1.29.2 (succeeds)
2. Test 2 = v2.3.0 (fails)
3. Test 3 = vv2.3.2-4-gd8775c7a (succeeds)

Test structure is:

1. Bring up container.
2. Show `/dev/video0` behaviour outside the container (proving camera is accessible).
3. Repeat same test inside the container - failure always expected because `/dev/video0` is not yet mapped.
4. Map `/dev/video0` from inside the container using `mknod`.
5. Show internal and external device mappings are the same.
6. Repeat test inside the container - success expected.
7. Terminate the container.

Critical lines:

* In test 2, "od: /dev/video0: Operation not permitted" shows step 4 did not permit access to `/dev/video0` from inside the container.
* In tests 1 and 3, successful octal dump shows access to `/dev/video0` was permitted.

```
Test 1
docker-compose version 1.29.2, build unknown
docker-py version: 5.0.3
CPython version: 3.7.3
OpenSSL version: OpenSSL 1.1.1d  10 Sep 2019
$ docker-compose up -d as-rule
Creating network "testing_default" with the default driver
Creating as-rule ... done
Check video device outside container:
$ od -t a -N 16 /dev/video0 | head -1
0000000 del   X del   a   c   f   E   x   i   f nul nul   M   M nul   *
Check video device inside container (failure expected):
$ docker exec as-rule bash -c "od -t a -N 16 /dev/video0 | head -1"
od: /dev/video0: No such file or directory
Map video device into container:
$ docker exec as-rule mknod /dev/video0 c 81 7
$ docker exec as-rule chgrp video /dev/video0
Show video device mapped into container:
$ ls -l /dev/video0 ; docker exec as-rule bash -c "ls -l /dev/video0"
crw-rw---- 1 root video 81, 7 Mar  9 09:58 /dev/video0
crw-r--r-- 1 root video 81, 7 Mar  9 11:55 /dev/video0
Re-check video device inside container (success expected):
$ docker exec as-rule bash -c "od -t a -N 16 /dev/video0 | head -1"
0000000 del   X del   a   c   f   E   x   i   f nul nul   M   M nul   *
Terminate test:
$ docker-compose down
Stopping as-rule ... done
Removing as-rule ... done
Removing network testing_default

Test 2
Docker Compose version v2.3.0
$ docker compose up -d as-rule
[+] Running 2/2
 ⠿ Network testing_default  Created                                                                                    0.5s
 ⠿ Container as-rule        Started                                                                                    1.3s
Check video device outside container:
$ od -t a -N 16 /dev/video0 | head -1
0000000 del   X del   a   c   f   E   x   i   f nul nul   M   M nul   *
Check video device inside container (failure expected):
$ docker exec as-rule bash -c "od -t a -N 16 /dev/video0 | head -1"
od: /dev/video0: No such file or directory
Map video device into container:
$ docker exec as-rule mknod /dev/video0 c 81 7
$ docker exec as-rule chgrp video /dev/video0
Show video device mapped into container:
$ ls -l /dev/video0 ; docker exec as-rule bash -c "ls -l /dev/video0"
crw-rw---- 1 root video 81, 7 Mar  9 09:58 /dev/video0
crw-r--r-- 1 root video 81, 7 Mar  9 11:55 /dev/video0
Re-check video device inside container (success expected):
$ docker exec as-rule bash -c "od -t a -N 16 /dev/video0 | head -1"
od: /dev/video0: Operation not permitted
Terminate test:
$ docker compose down
[+] Running 2/2
 ⠿ Container as-rule        Removed                                                                                   10.5s
 ⠿ Network testing_default  Removed                                                                                    0.4s

Test 3
Docker Compose version v2.3.2-4-gd8775c7a
$ /home/pi/compose/bin/docker-compose up -d as-rule
[+] Running 2/2
 ⠿ Network testing_default  Created                                                                                    0.5s
 ⠿ Container as-rule        Started                                                                                    1.2s
Check video device outside container:
$ od -t a -N 16 /dev/video0 | head -1
0000000 del   X del   a   c   f   E   x   i   f nul nul   M   M nul   *
Check video device inside container (failure expected):
$ docker exec as-rule bash -c "od -t a -N 16 /dev/video0 | head -1"
od: /dev/video0: No such file or directory
Map video device into container:
$ docker exec as-rule mknod /dev/video0 c 81 7
$ docker exec as-rule chgrp video /dev/video0
Show video device mapped into container:
$ ls -l /dev/video0 ; docker exec as-rule bash -c "ls -l /dev/video0"
crw-rw---- 1 root video 81, 7 Mar  9 09:58 /dev/video0
crw-r--r-- 1 root video 81, 7 Mar  9 11:55 /dev/video0
Re-check video device inside container (success expected):
$ docker exec as-rule bash -c "od -t a -N 16 /dev/video0 | head -1"
0000000 del   X del   a   c   f   E   x   i   f nul nul   M   M nul   *
Terminate test:
$ /home/pi/compose/bin/docker-compose down
[+] Running 2/2
 ⠿ Container as-rule        Removed                                                                                   10.5s
 ⠿ Network testing_default  Removed                                                                                    0.5s
```

**Comment by Paraphraser** [comment]
Seeing as discussion on this topic seems to be converging around this issue, I'll take the opportunity to cross-link to some others I found:

* #5622
* #5718
* #8251

I get the impression that `device_cgroup_rules` support fell out once before and had to be re-added.

**Comment by KevinRohn** [comment]
Thank you guys 👍
